File: clients/Ruby/model/entity.rb
require_relative 'entity_type'
require_relative 'vec2_int'
class Entity
    attr_accessor :id
    attr_accessor :player_id
    attr_accessor :entity_type
    attr_accessor :position
    attr_accessor :health
    attr_accessor :active
    def initialize(id, player_id, entity_type, position, health, active)
        @id = id
        @player_id = player_id
        @entity_type = entity_type
        @position = position
        @health = health
        @active = active
    end
    def self.read_from(stream)
        id = stream.read_int()
        if stream.read_bool()
            player_id = stream.read_int()
        else
            player_id = nil
        end
        entity_type = stream.read_int()
        if entity_type < 0 || entity_type > 10
            raise "Unexpected tag value"
        end
        position = Vec2Int.read_from(stream)
        health = stream.read_int()
        active = stream.read_bool()
        Entity.new(id, player_id, entity_type, position, health, active)
    end
    def write_to(stream)
        stream.write_int(@id)
        if @player_id.nil?
            stream.write_bool(false)
        else
            stream.write_bool(true)
            stream.write_int(@player_id)
        end
        stream.write_int(@entity_type)
        @position.write_to(stream)
        stream.write_int(@health)
        stream.write_bool(@active)
    end
end

File: clients/Cpp/model/Entity.cpp
#include "Entity.hpp"

Entity::Entity() { }
Entity::Entity(int id, std::shared_ptr<int> playerId, EntityType entityType, Vec2Int position, int health, bool active) : id(id), playerId(playerId), entityType(entityType), position(position), health(health), active(active) { }
Entity Entity::readFrom(InputStream& stream) {
    Entity result;
    result.id = stream.readInt();
    if (stream.readBool()) {
        result.playerId = std::shared_ptr<int>(new int());
        *result.playerId = stream.readInt();
    } else {
        result.playerId = std::shared_ptr<int>();
    }
    switch (stream.readInt()) {
    case 0:
        result.entityType = EntityType::WALL;
        break;
    case 1:
        result.entityType = EntityType::HOUSE;
        break;
    case 2:
        result.entityType = EntityType::BUILDER_BASE;
        break;
    case 3:
        result.entityType = EntityType::BUILDER_UNIT;
        break;
    case 4:
        result.entityType = EntityType::MELEE_BASE;
        break;
    case 5:
        result.entityType = EntityType::MELEE_UNIT;
        break;
    case 6:
        result.entityType = EntityType::RANGED_BASE;
        break;
    case 7:
        result.entityType = EntityType::RANGED_UNIT;
        break;
    case 8:
        result.entityType = EntityType::RESOURCE;
        break;
    case 9:
        result.entityType = EntityType::TURRET;
        break;
    default:
        throw std::runtime_error("Unexpected tag value");
    }
    result.position = Vec2Int::readFrom(stream);
    result.health = stream.readInt();
    result.active = stream.readBool();
    return result;
}
void Entity::writeTo(OutputStream& stream) const {
    stream.write(id);
    if (playerId) {
        stream.write(true);
        stream.write((*playerId));
    } else {
        stream.write(false);
    }
    stream.write((int)(entityType));
    position.writeTo(stream);
    stream.write(health);
    stream.write(active);
}
